### PR TITLE
Getting rid of episodes with activity and fixing episode counts

### PR DESCRIPTION
**Pull request by RobinGuillard**
- [X] Suppresses the episodes of MEMA/bruxism that overlap with unvalid labels from valid labels (OMA, wake, impedance checking issue). On the following image, it means keeping the first episode flageed "2" and getting rid of the 2nd.
image

- [X] Fixed issue that was present and now fixed on the decounting of the number of episodes:
image

- [X] Added the "labels_1s_extension" in utils function that enables to enlarge/extend OMA events

- [x] recomputed all results

### CODE REVIEW

**Comment by RobinGuillard** [comment]
Tried to run all subjects on Bruxism + MEMA and got the following error for 3NS01

```
3NS01_cohort2.edf (sleep labels 

WARNING:tinnsleep.data:non uniform interval (count: [(13.0, 1), (22.0, 1), (30.0, 1425)]), taking median: 30.0
WARNING:tinnsleep.data:start time, file: 22:13:38 labels: 25/08/2020 22:13:38

, loaded) preprocess... Bruxism(done) MEMA(done) DONE (44.34s) report... Bruxism(

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~\Documents\acou_sommeil_HD_ENS\Tinnitus-n-Sleep\compute_results.py in <module>
    263                                                       delim=delim, n_adaptive=n_adaptive_bruxism, log=log,
    264                                                       generate_report=generate_bruxism_report,
--> 265                                                       sleep_labels=sleep_labels_bruxism)
    266                     print(f"done)", end=" ")
    267                 else:

~\Documents\acou_sommeil_HD_ENS\Tinnitus-n-Sleep\tinnsleep\reports.py in reporting(epochs, valid_labels, THR_classif, time_interval, delim, n_adaptive, log, generate_report, sleep_labels, valid_sleep_stages)
    313         # -----------------Classification Forward ------------------------------------------------
    314         pipeline = AmplitudeThresholding(abs_threshold=THR[0], rel_threshold=THR[1], n_adaptive=abs(n_adaptive))
--> 315         labels = pipeline.fit_predict(X)
    316         if n_adaptive < -1:
    317             # -----------------Classification backward ---------------------------------------

~\Documents\acou_sommeil_HD_ENS\Tinnitus-n-Sleep\tinnsleep\classification.py in fit_predict(self, X, y)
    142     def fit_predict(self, X, y=[]):
    143         """Fit and predict in one function."""
--> 144         self.fit(X, y)
    145         return self.predict(X)

~\Documents\acou_sommeil_HD_ENS\Tinnitus-n-Sleep\tinnsleep\classification.py in fit(self, X, y, sample_weight)
     73 
     74         # Check that X and y have correct shape
---> 75         X = check_array(X)
     76 
     77         self.center_ = np.average(X, axis=0, weights=sample_weight)

~\Anaconda3\lib\site-packages\sklearn\utils\validation.py in check_array(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, warn_on_dtype, estimator)
    548                              " minimum of %d is required%s."
    549                              % (n_samples, array.shape, ensure_min_samples,
--> 550                                 context))
    551 
    552     if ensure_min_features > 0 and array.ndim == 2:

ValueError: Found array with 0 sample(s) (shape=(0, 1)) while a minimum of 1 is required.
```

**Comment by lkorczowski** [comment]
> Tried to run all subjects on Bruxism + MEMA and got the following error for 3NS01
> 
> ```
> 3NS01_cohort2.edf (sleep labels 
> 
> WARNING:tinnsleep.data:non uniform interval (count: [(13.0, 1), (22.0, 1), (30.0, 1425)]), taking median: 30.0
> WARNING:tinnsleep.data:start time, file: 22:13:38 labels: 25/08/2020 22:13:38
> 
> , loaded) preprocess... Bruxism(done) MEMA(done) DONE (44.34s) report... Bruxism(
> 
> ---------------------------------------------------------------------------
> ValueError                                Traceback (most recent call last)
> ~\Documents\acou_sommeil_HD_ENS\Tinnitus-n-Sleep\compute_results.py in <module>
>     263                                                       delim=delim, n_adaptive=n_adaptive_bruxism, log=log,
>     264                                                       generate_report=generate_bruxism_report,
> --> 265                                                       sleep_labels=sleep_labels_bruxism)
>     266                     print(f"done)", end=" ")
>     267                 else:
> 
> ~\Documents\acou_sommeil_HD_ENS\Tinnitus-n-Sleep\tinnsleep\reports.py in reporting(epochs, valid_labels, THR_classif, time_interval, delim, n_adaptive, log, generate_report, sleep_labels, valid_sleep_stages)
>     313         # -----------------Classification Forward ------------------------------------------------
>     314         pipeline = AmplitudeThresholding(abs_threshold=THR[0], rel_threshold=THR[1], n_adaptive=abs(n_adaptive))
> --> 315         labels = pipeline.fit_predict(X)
>     316         if n_adaptive < -1:
>     317             # -----------------Classification backward ---------------------------------------
> 
> ~\Documents\acou_sommeil_HD_ENS\Tinnitus-n-Sleep\tinnsleep\classification.py in fit_predict(self, X, y)
>     142     def fit_predict(self, X, y=[]):
>     143         """Fit and predict in one function."""
> --> 144         self.fit(X, y)
>     145         return self.predict(X)
> 
> ~\Documents\acou_sommeil_HD_ENS\Tinnitus-n-Sleep\tinnsleep\classification.py in fit(self, X, y, sample_weight)
>      73 
>      74         # Check that X and y have correct shape
> ---> 75         X = check_array(X)
>      76 
>      77         self.center_ = np.average(X, axis=0, weights=sample_weight)
> 
> ~\Anaconda3\lib\site-packages\sklearn\utils\validation.py in check_array(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, warn_on_dtype, estimator)
>     548                              " minimum of %d is required%s."
>     549                              % (n_samples, array.shape, ensure_min_samples,
> --> 550                                 context))
>     551 
>     552     if ensure_min_features > 0 and array.ndim == 2:
> 
> ValueError: Found array with 0 sample(s) (shape=(0, 1)) while a minimum of 1 is required.
> ```

`check_array` did the job well. X is an empty array of shape (0,1)

I will check it out tomorrow. 

Meanwhile : #107 fix the issue of having to remove any file from the analysis.

File: tinnsleep/reports.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 from tinnsleep.signal import is_good_epochs, power_ratio
 
 
-def generate_bruxism_report(classif, time_interval, delim, min_burst_joining=3, sleep_labels=None):
+def generate_bruxism_report(classif, time_interval, delim, valid_labels, min_burst_joining=3, sleep_labels=None):
```
Comment (lkorczowski) [inline tinnsleep/reports.py]: valid_labels shouldn't be a mandatory parameters (I'm dealing with it).

File: tinnsleep/reports.py [hunk 2]
```diff
@@ -98,11 +114,20 @@ def generate_MEMA_report(classif, time_interval, delim, sleep_labels=None):
...
     nb_burst = len(li_burst)
     report["Total number of MEMA burst"] = nb_burst
     report["Number of MEMA bursts per hour"] = nb_burst * 3600 / recording_duration
     li_episodes = burst_to_episode(li_burst, delim=delim, min_burst_joining=0)
+    nb_episodes = len(li_episodes)
+
+    new_ep_list = []
+    for ep in li_episodes:
+        if np.sum(np.invert(valid_labels[int(ep.beg / time_interval):int(ep.end / time_interval)])) == 0:
+            new_ep_list.append(ep)
```
Comment (lkorczowski) [inline tinnsleep/reports.py:138]: smart !

File: tinnsleep/utils.py [hunk 3]
```diff
@@ -150,7 +150,8 @@ def fuse_with_classif_result(check_imp, labels):
     mod_labels : ndarray, shape (n_trials,)
         Modified labels to fit a length of n_trials
     """
-    labels = labels.tolist()
+    if not isinstance(labels, list):
+        labels = labels.tolist()
```
Comment (lkorczowski) [inline tinnsleep/utils.py:154]: nice catch

File: tinnsleep/utils.py [hunk 4]
```diff
@@ -150,7 +150,8 @@ def fuse_with_classif_result(check_imp, labels):
...
     for i in range(len(check_imp)):
         if np.mean(check_imp[i]) == 1:
             labels.insert(i, False)  # inserting False (neutral) values in the labels array where they were deleted
```
Comment (lkorczowski) [inline tinnsleep/utils.py:157]: now that I think about it, the use of `np.mean` is not great.

Also, the insert in labels shouldnt work as expected (insert will shift all index so "i" is not the good reference). I need to check how it has been pytested.

File: tinnsleep/utils.py [hunk 5]
```diff
@@ -295,3 +296,94 @@ def round_time(dt=None, round_to=60):
...
+                flag = False
+        else:
+            if flag == False:
+                if elm == 0:
+                    new_l.append(0)
+                else:
+                    new_l[-1] = 1
+                    new_l.append(1)
+                    flag = True
+    return new_l
```
Comment (lkorczowski) [inline tinnsleep/utils.py]: omg. I will give you the medal of "master of lists". I need to double check carefully that :/ 

XD

File: tinnsleep/utils.py [hunk 6]
```diff
@@ -295,3 +296,94 @@ def round_time(dt=None, round_to=60):
...
+    -------
+    new_l : list of booleans/ 0s and 1s of the same length as l
+        the same list where every time a series of 1 is present, it is completed by added_left 1s on the left and
+        added_right 1s on the right
+    """
+    for i in range(added_left):
+        l = extending_episodes_left(l)
+    for i in range(added_right):
+        l = extending_episodes_right(l)
+    return l
```
Comment (lkorczowski) [inline tinnsleep/utils.py]: OK got it !
 I believe the aim is to expend the artifacts tags. A description would be great.

File: tinnsleep/utils.py (hunk 5 above)
Comment (lkorczowski) [reply tinnsleep/utils.py]: ok, so I reviewed that code and I see less and less its purpose, so my comment about using "+" instead.

Please explain the use case.

File: tinnsleep/utils.py [hunk 7]
```diff
@@ -295,3 +296,94 @@ def round_time(dt=None, round_to=60):
...
+                if elm == 0:
+                    new_l.append(0)
+                else:
+                    new_l[-1] = 1
+                    new_l.append(1)
+                    flag = True
+    return new_l
+
+def labels_1s_extension(l, added_left, added_right):
+    """
```
Comment (lkorczowski) [inline tinnsleep/utils.py:372]: description and use case missing

File: tinnsleep/reports.py [hunk 8]
```diff
@@ -28,12 +32,20 @@ def generate_bruxism_report(classif, time_interval, delim, min_burst_joining=3,
     recording_duration = len(classif) * time_interval
     report["Clean data duration"] = recording_duration
     report["Total burst duration"] = np.sum(classif) * time_interval
-    li_burst = classif_to_burst(classif, time_interval)
+    classif_good_length = fuse_with_classif_result(np.invert(valid_labels),
```
Comment (lkorczowski) [inline tinnsleep/reports.py]: Carefull if you use `np.invert` as:
```
np.invert([1,0])  # matrix inversion
> array([-2, -1]) 
```

fix:
```
np.invert(np.array([0,1])==1)
> array([ True, False])
```

Comment (lkorczowski) [reply tinnsleep/reports.py]: it is done. (see new version)

File: tinnsleep/reports.py [hunk 9]
```diff
@@ -98,11 +114,20 @@ def generate_MEMA_report(classif, time_interval, delim, sleep_labels=None):
     recording_duration = len(classif) * time_interval
     report["Clean MEMA duration"] = recording_duration
     report["Total MEMA burst duration"] = np.sum(classif) * time_interval
-    li_burst = classif_to_burst(classif, time_interval=time_interval)
+    classif_good_length = fuse_with_classif_result(np.invert(valid_labels),
```
Comment (lkorczowski) [inline tinnsleep/reports.py]: same problem (dealt with, see new lines)

File: tinnsleep/utils.py (hunk 3 above)
Comment (lkorczowski) [reply tinnsleep/utils.py:154]: but the fact that it uses list only is suboptimal (numpy is almost always better)

I'm creating an issue to refactor this funciton

**Review by lkorczowski — CHANGES_REQUESTED** [review]
Here are some keys issues:
- [x] bad naming for variable and functions (mostly fixed by commit)
- [x] issue with valid_labels in generate_report: `np.invert` could have been a problem. This issue is not completely fix. I feel that something could be wrong and we need more pytesting I think.
- [ ] NB check: not done yet. Will do after everything else.

Please read the comment. If the comment is noted "obsolete" it is because I try to fix.

File: tinnsleep/reports.py (hunk 1 above)
Comment (lkorczowski) [reply tinnsleep/reports.py]: done

**Comment by lkorczowski** [comment]
@RobinGuillard : PR done. We should merge to master asap for fixing other stuff in smaller batch.
Note: I didn't check the NB, should I ?

File: tinnsleep/utils.py (hunk 5 above)
Comment (RobinGuillard) [reply tinnsleep/utils.py]: I believe you found your answer just below!

File: tinnsleep/utils.py (hunk 6 above)
Comment (RobinGuillard) [reply tinnsleep/utils.py]: Yep!
Ok for the description...

File: tinnsleep/utils.py (hunk 7 above)
Comment (RobinGuillard) [reply tinnsleep/utils.py:372]: Ok ok

File: tinnsleep/reports.py (hunk 8 above)
Comment (RobinGuillard) [reply tinnsleep/reports.py]: Great!

**Pull request merged**